File: token/token.go
// Package token defines the lexical tokens of the Iris language.
package token

import "fmt"

// Kind identifies the category of a token.
type Kind int

const (
	EOF     Kind = iota
	ILLEGAL      // unexpected character

	IDENT  // identifier: foo, Bar, myVar
	INT    // integer literal: 42
	FLOAT  // float literal: 3.14
	STRING // string literal: "hello"

	// ── Keywords ────────────────────────────────────────────────────────────
	KW_LET    // let
	KW_FN     // fn
	KW_TYPE   // type
	KW_TRAIT  // trait
	KW_IMPL   // impl
	KW_USE    // use
	KW_FOR    // for
	KW_EXPORT // export
	KW_MATCH  // match
	KW_TRUE   // true
	KW_FALSE  // false

	// ── Backbone pipeline operators ──────────────────────────────────────────
	// All backbone ops are :keyword or :: form
	OP_NEXT        // ::     continue (sync)
	OP_IF          // :if    conditional branch
	OP_THEN        // :then  branch body (used by :if and :while)
	OP_ELSE        // :else  fallback branch
	OP_WHILE       // :while loop condition
	OP_CATCH_NAMED // :catch named-error catch
	OP_AWAIT       // :await async wait
	OP_OR          // :or    fallback value
	OP_TRY         // :try   propagate error (short-circuit)
	OP_TAG         // :tag   borrow-save snapshot
	OP_JOIN        // :join  tuple join (subject, tag)

	// ── Arithmetic / relational operators ───────────────────────────────────
	PLUS    // +
	MINUS   // -
	STAR    // *
	SLASH   // /
	PERCENT // %

	EQEQ   // ==
	NEQ    // !=
	LT     // <
	GT     // >
	LTEQ   // <=
	GTEQ   // >=
	ANDAND // &&
	OROR   // ||
	ARROW  // ->

	// ── Delimiters ───────────────────────────────────────────────────────────
	LPAREN   // (
	RPAREN   // )
	LBRACE   // {
	RBRACE   // }
	LBRACKET // [
	RBRACKET // ]

	// ── Punctuation ──────────────────────────────────────────────────────────
	COMMA     // ,
	DOT       // .
	COLON     // :
	SEMICOLON // ;
	EQUAL     // =  (standalone, not part of =:)

	// ── Iris-specific sigils ─────────────────────────────────────────────────
	HASH   // #  trait reference prefix
	DOLLAR // $  subject position / alias
	BANG   // !  comptime directive prefix
	AT     // @  decorator
	PIPE   // |  join branch separator
	AMP    // &
	TILDE  // ~
	CARET  // ^
	QUEST  // ?
)

var keywords = map[string]Kind{
	"let":    KW_LET,
	"fn":     KW_FN,
	"type":   KW_TYPE,
	"trait":  KW_TRAIT,
	"impl":   KW_IMPL,
	"use":    KW_USE,
	"for":    KW_FOR,
	"export": KW_EXPORT,
	"match":  KW_MATCH,
	"true":   KW_TRUE,
	"false":  KW_FALSE,
}

// LookupIdent returns the keyword Kind for s, or IDENT if s is not a keyword.
func LookupIdent(s string) Kind {
	if k, ok := keywords[s]; ok {
		return k
	}
	return IDENT
}

var kindName = map[Kind]string{
	EOF: "EOF", ILLEGAL: "ILLEGAL",
	IDENT: "IDENT", INT: "INT", FLOAT: "FLOAT", STRING: "STRING",
	KW_LET: "let", KW_FN: "fn", KW_TYPE: "type", KW_TRAIT: "trait",
	KW_IMPL: "impl", KW_USE: "use", KW_FOR: "for", KW_EXPORT: "export", KW_MATCH: "match",
	KW_TRUE: "true", KW_FALSE: "false",
	OP_NEXT: "::", OP_IF: ":if", OP_THEN: ":then", OP_ELSE: ":else",
	OP_WHILE: ":while", OP_CATCH_NAMED: ":catch", OP_AWAIT: ":await",
	OP_OR: ":or", OP_TRY: ":try", OP_TAG: ":tag", OP_JOIN: ":join",
	PLUS: "+", MINUS: "-", STAR: "*", SLASH: "/", PERCENT: "%",
	EQEQ: "==", NEQ: "!=", LT: "<", GT: ">", LTEQ: "<=", GTEQ: ">=",
	ANDAND: "&&", OROR: "||", ARROW: "->",
	LPAREN: "(", RPAREN: ")", LBRACE: "{", RBRACE: "}", LBRACKET: "[", RBRACKET: "]",
	COMMA: ",", DOT: ".", COLON: ":", SEMICOLON: ";", EQUAL: "=",
	HASH: "#", DOLLAR: "$", BANG: "!", AT: "@", PIPE: "|", AMP: "&",
	TILDE: "~", CARET: "^", QUEST: "?",
}

func (k Kind) String() string {
	if s, ok := kindName[k]; ok {
		return s
	}
	return fmt.Sprintf("Kind(%d)", int(k))
}

// IsBackboneOp reports whether k is a pipeline backbone operator.
func (k Kind) IsBackboneOp() bool {
	return k >= OP_NEXT && k <= OP_JOIN
}

// Token is a single lexical unit produced by the lexer.
type Token struct {
	Kind   Kind
	Lexeme string // raw text from source
	Line   int
	Col    int
}

func (t Token) String() string {
	return fmt.Sprintf("%v %q (%d:%d)", t.Kind, t.Lexeme, t.Line, t.Col)
}

File: src/token.rs
//! 字句解析が生成するトークン。各トークンは種別と span を持つ。

use crate::span::Span;

#[derive(Debug, Clone, PartialEq)]
pub enum TokenKind {
    // リテラル
    Int(i64),
    Float(f64),
    Str(String),
    Bool(bool),
    Ident(String),

    // キーワード
    Fn,
    Let,
    Const,
    Return,
    If,
    Else,
    While,
    Loop,
    Break,
    Continue,
    Pub,
    Mut,
    Type,
    Struct,
    Enum,
    Extern,
    Match,

    // 区切り・括弧
    LParen,   // (
    RParen,   // )
    LBrace,   // {
    RBrace,   // }
    LBracket, // [
    RBracket, // ]
    Comma,    // ,
    Colon,    // :
    Dot,      // .
    DotDot,   // ..  (範囲パターン・排他上限)
    DotDotEq, // ..= (範囲パターン・包含上限)
    Arrow,    // ->

    // 演算子
    Plus,     // +
    Minus,    // -
    Star,     // *
    Slash,    // /
    Percent,  // %
    Assign,   // =
    EqEq,     // ==
    NotEq,    // !=
    Lt,       // <
    LtEq,     // <=
    Gt,       // >
    GtEq,     // >=
    AndAnd,   // &&
    OrOr,     // ||
    Bang,     // !  (エラー伝播の後置演算子)
    Question, // ?  (三項演算子)
    Amp,      // &  (参照)
    Hash,     // #  (メソッド名衝突の修飾子。ADR-0004)

    // 構造
    Newline,
    Eof,
}

impl TokenKind {
    /// 診断メッセージ向けの人間可読な名前。
    pub fn describe(&self) -> String {
        match self {
            TokenKind::Int(_) => "整数リテラル".into(),
            TokenKind::Float(_) => "浮動小数点リテラル".into(),
            TokenKind::Str(_) => "文字列リテラル".into(),
            TokenKind::Bool(_) => "真偽値リテラル".into(),
            TokenKind::Ident(_) => "識別子".into(),
            TokenKind::Fn => "`fn`".into(),
            TokenKind::Let => "`let`".into(),
            TokenKind::Const => "`const`".into(),
            TokenKind::Return => "`return`".into(),
            TokenKind::If => "`if`".into(),
            TokenKind::Else => "`else`".into(),
            TokenKind::While => "`while`".into(),
            TokenKind::Loop => "`loop`".into(),
            TokenKind::Break => "`break`".into(),
            TokenKind::Continue => "`continue`".into(),
            TokenKind::Pub => "`pub`".into(),
            TokenKind::Mut => "`mut`".into(),
            TokenKind::Type => "`type`".into(),
            TokenKind::Struct => "`struct`".into(),
            TokenKind::Enum => "`enum`".into(),
            TokenKind::Extern => "`extern`".into(),
            TokenKind::Match => "`match`".into(),
            TokenKind::LParen => "`(`".into(),
            TokenKind::RParen => "`)`".into(),
            TokenKind::LBrace => "`{`".into(),
            TokenKind::RBrace => "`}`".into(),
            TokenKind::LBracket => "`[`".into(),
            TokenKind::RBracket => "`]`".into(),
            TokenKind::Comma => "`,`".into(),
            TokenKind::Colon => "`:`".into(),
            TokenKind::Dot => "`.`".into(),
            TokenKind::DotDot => "`..`".into(),
            TokenKind::DotDotEq => "`..=`".into(),
            TokenKind::Arrow => "`->`".into(),
            TokenKind::Plus => "`+`".into(),
            TokenKind::Minus => "`-`".into(),
            TokenKind::Star => "`*`".into(),
            TokenKind::Slash => "`/`".into(),
            TokenKind::Percent => "`%`".into(),
            TokenKind::Assign => "`=`".into(),
            TokenKind::EqEq => "`==`".into(),
            TokenKind::NotEq => "`!=`".into(),
            TokenKind::Lt => "`<`".into(),
            TokenKind::LtEq => "`<=`".into(),
            TokenKind::Gt => "`>`".into(),
            TokenKind::GtEq => "`>=`".into(),
            TokenKind::AndAnd => "`&&`".into(),
            TokenKind::OrOr => "`||`".into(),
            TokenKind::Bang => "`!`".into(),
            TokenKind::Question => "`?`".into(),
            TokenKind::Amp => "`&`".into(),
            TokenKind::Hash => "`#`".into(),
            TokenKind::Newline => "改行".into(),
            TokenKind::Eof => "入力の終端".into(),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub kind: TokenKind,
    pub span: Span,
}

impl Token {
    pub fn new(kind: TokenKind, span: Span) -> Self {
        Token { kind, span }
    }
}
